### DISCUSSION TITLE
Working with monomial vectors

### DISCUSSION THREAD

**Discussion by Fabian-Geyer** [iFR-ACSO/casos#78] [General]
Before introducing the sparsity concept for the monomial vectors, working with them could sometimes be more intuitive (from a beginners perspective). Consider the literature example from [1] where one could write:

```matlab
x = casos.PS.sym('x', 2);

% create monomials vectors
z_1 = monomials(x, [1:1]);
z_2 = monomials(x, 2:2);
z_3 = monomials(x, 3:3);

% dynamics
f = A_1*z_1 + A_2*z_2 + A_3*z_3;
```
It might not be "best practice", but is there another intuitive way of obtaining something similar with the newer version?

@tcunis / @renatoloureiro 
[1] P. Seiler and G. J. Balas, “Quasiconvex sum-of-squares programming,” in 49th IEEE Conference on Decision and Control (CDC), Atlanta, GA, USA: IEEE, Dec. 2010, pp. 3337–3342. doi: 10.1109/CDC.2010.5717672.

**Comment by JOlucak** [comment]
One possible way is:

% generate random polynomial
x = casos.PS('x',2);
mon = monomials(x,0:2);
a    = rand(mon.nnz,1)
poly = casos.PS(,a)


In the above case

% create monomials vectors
z_1 = monomials(x, [1:1]);
z_2 = monomials(x, 2:2);
z_3 = monomials(x, 3:3);

% dynamics
A1 = rand(z_1.nnz,1)
A2 = rand(z_2.nnz,1)
A3 = rand(z_3.nnz,1)

% f = A_1*z_1 + A_2*z_2 + A_3*z_3;
f = casos.PS(z_1,A1) + casos.PS(z_2,A2) + casos.PS(z_3,A3)

**Reply by tcunis** [reply]
this works provided that `A1` through `A3` are vectors which, in the example of Seiler and Balas, they're not.

**Comment by tcunis** [comment]
It is important to note that the monomial sparsity pattern is fundamentally different from a vector of monomials. The latter is a vector-valued polynomial which has an onto-relation between monomials and entries (often in a sorted manner, i.e., $z_\alpha = x^\alpha$). The monomial sparsity pattern describe the monomials in a compact form and is, here, scalar-valued. 

In sosopt, the `monomials` method is used to create a vector of monomials as sosopt does not have a sparsity concept. I have kept the syntax but return a monomial sparsity pattern; this is motivated from the following use case:

```
z = monomials(x, degs);
s = casos.PS.sym('q',z);
```

In fact, the `monomials` method is internally mapped to `casos.Sparsity.scalar`.

**Reply by tcunis** [reply]
currently, the only way to implement the example is to use the `to_vector` method for sparsity patterns, then rely on the implicit casting from `casos.Sparsity` to `casos.PD`. This effectively converts a monomial sparsity pattern into a vector of monomials:

```
z1 = to_vector(monomials(x,1));
...
z3 = to_vector(monomials(x,3));
```

However, observe that the variables `z1` through `z3` are still of type `casos.Sparsity` and the implicit casting might not be desirable.

**Reply by tcunis** [reply]
Alternatively, I suggest to overload the `mpower` operator for indeterminate variables to return a vector of monomials (of type `casos.PD`). Then, the example could look like

```
f = A1*x + A2*x^2 + A3*x^3;
```
